examples/peft_llm_gnn/llm_gnn_model.py

```diff
@@ -192,10 +221,169 @@ def create_optimizer(self):
+        prompt = self.encode_graph(blocks, h)
+        if prompt is None:
+            model_output = self.llm(input_ids, attention_mask)
+            masked_hidden_states = model_output.last_hidden_state * attention_mask.unsqueeze(-1)
+            last_token_indexes = (attention_mask.sum(dim=1, dtype=torch.int64) - 1)
+            last_token_embeddings = masked_hidden_states[torch.arange(last_token_indexes.size(0)),last_token_indexes,:]
+        else:
+            word_embeddings = self.llm.get_input_embeddings()
+            # assuming graph_tokens has a shape of [batch_size, embedding_dim],
+            # input_ids has a shape of [batch_size, seq_len, embedding_dim]
+            # only one graph token is inserted ahead of input words 
+            inputs_embeds = torch.cat([prompt.unsqueeze(1), word_embeddings(input_ids)], dim=1)
+            attention_mask = torch.cat([torch.ones((input_ids.shape[0],1), device=self.llm.device), attention_mask], dim=1)
+            model_output = self.llm(inputs_embeds=inputs_embeds, attention_mask=attention_mask)
+            
+            masked_hidden_states = model_output.last_hidden_state * attention_mask.unsqueeze(-1)
+            last_token_indexes = (attention_mask.sum(dim=1, dtype=torch.int64) - 1)
+            last_token_embeddings = masked_hidden_states[torch.arange(last_token_indexes.size(0)),last_token_indexes,:]
+        return {self.target_ntype: F.normalize(last_token_embeddings, p=2)}
+
+class GNNLLM_LP(gsmodel.GSgnnLinkPredictionModelBase):
+    """ GNNLLM_LP Model Class
+
+    Parameters
+    ----------
+    g : dgl DistGraph
+        Input DistDGL graph
+    node_lm_configs : list
+        language model config for each node type
+    h_dim : int
+        hidden dimension of GNN encoder
+    out_dim : int
+        Output dimension of the model, e.g. number of classes.
+    num_layers : int
+        Number of GNN encoder layers.
+    target_ntype : str
+        The node type for prediction.
+    use_norm: boolean, optional
+        If use layer normalization or not. Default: True
+    alpha_l2norm: float, optional
+        The alpha for L2 normalization. Default: 0
+    lr : float, optional
+        Normalization Method. Default: 0.001
+    """
+    def __init__(self, g, node_lm_configs, h_dim, num_layers,
+                 target_ntype,     # the node type to be predict
+                 target_etype,     # the node type to be predict
+                 use_norm = True,   # use normalization or not, default is True
+                 alpha_l2norm = 0,
+                 lr = 0.001):
+        super(GNNLLM_LP, self).__init__()
+        
+        self.target_etype=target_etype
+        self.alpha_l2norm = alpha_l2norm
+        self.lr = lr
+        self.gnn_encoder = GPEFT(g, h_dim, num_layers, target_ntype, node_lm_configs)
+        # dummy node input encoder to enable LP Infer
+        self.node_input_encoder = DummyNodeInputLayer()
```

Comment:
do you have to set up this field?

Reply:
where is the field used? i searched for it. it seems this is the only place where it is used.

Reply:
It's here. If we don't have model.node_input_encoder, it will report exception. 
https://github.com/awslabs/graphstorm/blob/b38259de64877455f8dc965582a5dc180eda807d/python/graphstorm/model/gnn.py#L985C35-L985C54